### PR TITLE
Fixes 1963: nightly sync

### PR DESCRIPTION
**Pull request by dpang314**
## Summary

Adds repository snapshotting to nightly tasks. When the snapshot feature is enabled in the config, nightly tasks will create a snapshot of all repositories with snapshot = true that haven't been snapshotted in the last 24 hours.

## Testing steps

Enable the snapshot feature in the config and create a repository with snapshotting enabled. 
`go run cmd/external-repos/main.go nightly-jobs` triggers the sync task. 
If there is a repository with snapshotting enabled that hasn't been snapshotted in the past 24 hours (including one that has no snapshots), then a snapshot task for that repository will be enqueued. Running the nightly-job will log that the snapshot task was enqueued.

### CODE REVIEW

File: cmd/external-repos/main.go [hunk 1]
```diff
@@ -150,3 +157,38 @@ func enqueueIntrospectAllRepos() error {
...
+		return fmt.Errorf("error getting repository configurations: %w", err)
+	}
+
+	snapshotDao := dao.GetSnapshotDao(db.DB)
+	currentTime := time.Now()
+	for _, repo := range repoConfigs {
+		snapshot, err := snapshotDao.FetchLatestSnapshot(repo.UUID)
+		if err != nil && err != gorm.ErrRecordNotFound {
+			log.Err(err).Msgf("error fetching snapshot for repository %v", repo.Name)
+		} else if err == gorm.ErrRecordNotFound || currentTime.Sub(snapshot.CreatedAt) > time.Hour*24 {
```
[the remark this answers is not in the document]
Comment (dpang314) [reply cmd/external-repos/main.go:180]: Got it, makes sense.

File: pkg/dao/snapshots.go [hunk 2]
```diff
@@ -78,3 +84,19 @@ func (sDao snapshotDaoImpl) Delete(snapUUID string) error {
...
 }
+
+func (sDao snapshotDaoImpl) FetchLatestSnapshot(repoConfigUUID string) (api.SnapshotResponse, error) {
+	var snap models.Snapshot
+	result := sDao.db.
+		Joins("inner join repository_configurations on repository_configurations.repository_uuid = snapshots.repository_uuid").
+		Where("repository_configurations.uuid = ?", repoConfigUUID).
+		Where("snapshots.org_id = repository_configurations.org_id").
+		Order("created_at DESC").
+		First(&snap)
```
Comment (jlsherrill) [inline pkg/dao/snapshots.go:95]: Andrew's pr makes this a lot simpler by adding a latest_snapshot_uuid column to repo_config, but this is good for now

**Comment by swadeley** [comment]
Hi, is this ready for testing? Please rebase.

**Review by jlsherrill — APPROVED** [review]
Thanks @dpang314 !!

**Comment by jlsherrill** [comment]
feel free to qe this, but we need to probably hold off on merging until snapshotting is fixed in stage

**Comment by swadeley** [comment]
> feel free to qe this, but we need to probably hold off on merging until snapshotting is fixed in stage

HI, I can test this on Friday, so I can leave the test ephemeral env running overnight.

**Comment by swadeley** [comment]
Hi

test started:

```
In [1]:      repo = dict(
   ...:            snapshot=True,
   ...:            name='fedoratest-1',
   ...:            url='https://stephenw.fedorapeople.org/multirepos/5/repo01/'
   ...:       )

In [2]: app.content_sources.rest_client.repositories_api.create_repository(repo)
2023-08-17 21:54:54.336 [    INFO] [root] Using <function client_obj_maker.<locals>.make_obj at 0x7f1417021a60> object....with url https://env-ephemeral-jo9225-wi6oqp1z.apps.c-rh-c-eph.8p0c.p1.openshiftapps.com/api/content-sources/v1 and verify_ssl set to True
2023-08-17 21:54:54.337 [    INFO] [iqe.base.auth] Available auth types: ['basic', 'cert', 'identity', 'jwt', 'uht']
2023-08-17 21:54:54.337 [    INFO] [iqe.base.auth] Setting auth_type to jwt
2023-08-17 21:54:56.670 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=nZUgqDVxGZOSdJIhsAY6GH7W7kMrM1wn, params=[]
Out[2]: 
{'account_id': '0369233',
 'distribution_arch': 'any',
 'distribution_versions': ['any'],
 'failed_introspections_count': 0,
 'gpg_key': '',
 'last_introspection_error': '',
 'last_introspection_time': '',
 'last_success_introspection_time': '',
 'last_update_introspection_time': '',
 'metadata_verification': False,
 'name': 'fedoratest-1',
 'org_id': '3340851',
 'package_count': 0,
 'snapshot': True,
 'status': 'Pending',
 'url': 'https://stephenw.fedorapeople.org/multirepos/5/repo01/',
 'uuid': '5bad1ff3-2757-4b8b-8878-8947be7d8e96'}

In [3]: app.content_sources.rest_client.tasks_api.list_tasks()
2023-08-17 21:55:18.573 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=2hM2ZiIIIuRwT1TETyGInrOxtHT7ZM7R, params=[]
Out[3]: 
{'data': [{'created_at': '2023-08-17T20:54:56Z',
           'ended_at': '2023-08-17T20:54:56Z',
           'error': '',
           'org_id': '3340851',
           'status': 'completed',
           'uuid': '43016219-585f-4778-9ebb-3a50f980d0f3'},
          {'created_at': '2023-08-17T20:54:56Z',
           'ended_at': '2023-08-17T20:55:06Z',
           'error': '',
           'org_id': '3340851',
           'status': 'completed',
           'uuid': '95a44239-724c-4b3d-943b-0231a95f7f61'}],
 'links': {'first': '/api/content-sources/v1/tasks/?limit=100&offset=0',
           'last': '/api/content-sources/v1/tasks/?limit=100&offset=0'},
 'meta': {'count': 2, 'limit': 100, 'offset': 0}}

In [4]: 


```

**Comment by swadeley** [comment]
Hi,

I see repos that have been deleted being introspected, not sure yet if that happens just once. Asked in slack.

**Comment by swadeley** [comment]
Hi @Andrewgdewar 

As discussed, I was seeing introspection never ending while testing this.

Please rebase for me to pull in latest Pulp improvements.

Thank you

**Comment by swadeley** [comment]
Need to test snapshot is created every 24 hours if no snapshot was already made
   We can check  'last_snapshot' values and  'last_snapshot_uuid' 
    --------------------------------------------------------------------------------------------------------------------------------


     
    In [1]:      repo = dict(
       ...:            snapshot=True,
       ...:            name='pulp-core',
       ...:            url='https://yum.theforeman.org/pulpcore/3.4/el7/x86_64/'
       ...:       )
     
    In [2]: app.content_sources.rest_client.repositories_api.create_repository(repo)
    2023-09-04 19:21:59.837 [    INFO] [root] Using <function client_obj_maker.<locals>.make_obj at 0x7f05a9983940> object....with url https://env-ephemeral-wfqmgj-ek8gfkpo.apps.c-rh-c-eph.8p0c.p1.openshiftapps.com/api/content-sources/v1 and verify_ssl set to True
    2023-09-04 19:21:59.837 [    INFO] [iqe.base.auth] Available auth types: ['basic', 'cert', 'identity', 'jwt', 'uht']
    2023-09-04 19:21:59.837 [    INFO] [iqe.base.auth] Setting auth_type to jwt
    2023-09-04 19:22:01.075 [    INFO] [iqe.base.rest_client] REST: METHOD=POST, request_id=m47XLg3re2D3ZEwmSHMYVsehBRyU89Xo, params=[]
    Out[2]:
    {'account_id': '0369233',
     'distribution_arch': 'any',
     'distribution_versions': ['any'],
     'failed_introspections_count': 0,
     'gpg_key': '',
     'last_introspection_error': '',
     'last_introspection_time': '',
     '**last_snapshot_task_uuid**': '50e7f152-c65f-4591-8318-807762e204fc',
     'last_success_introspection_time': '',
     'last_update_introspection_time': '',
     'metadata_verification': False,
     'name': 'pulp-core',
     'org_id': '3340851',
     'package_count': 0,
     'snapshot': True,
     'status': 'Pending',
     'url': 'https://yum.theforeman.org/pulpcore/3.4/el7/x86_64/',
     'uuid': '5ccad0eb-56e3-4ba1-a76f-60842bb9899f'}
     
     
    In [1]: app.content_sources.rest_client.repositories_api.list_repositories()
    2023-09-05 13:43:38.486 [    INFO] [root] Using <function client_obj_maker.<locals>.make_obj at 0x7f0cf55e1790> object....with url https://env-ephemeral-wfqmgj-ek8gfkpo.apps.c-rh-c-eph.8p0c.p1.openshiftapps.com/api/content-sources/v1 and verify_ssl set to True
    2023-09-05 13:43:38.486 [    INFO] [iqe.base.auth] Available auth types: ['basic', 'cert', 'identity', 'jwt', 'uht']
    2023-09-05 13:43:38.486 [    INFO] [iqe.base.auth] Setting auth_type to jwt
    2023-09-05 13:43:39.427 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=qw1eHltXTG5HaP1AwXkvQsqbGKBzFVrG, params=[]
    Out[1]:
    {'data': [{'account_id': '0369233',
               'distribution_arch': 'any',
               'distribution_versions': ['any'],
               'failed_introspections_count': 0,
               'gpg_key': '',
               'last_introspection_error': '',
               'last_introspection_time': '2023-09-04T18:46:01Z',
               'last_snapshot': {'added_counts': {'rpm.package': 121},
                                 'content_counts': {'rpm.package': 121},
                                 '**created_at**': '2023-09-04T18:46:00.633897Z',
                                 'removed_counts': {},
                                 'repository_path': ''},
               '**last_snapshot_task_uuid**': '50e7f152-c65f-4591-8318-807762e204fc',
               '**last_snapshot_uuid**': '679906d4-091e-4d26-bf61-3586e3796a3c',
               'last_success_introspection_time': '2023-09-04T18:46:01Z',
               'last_update_introspection_time': '2023-09-04T18:46:01Z',
               'metadata_verification': False,
               'name': 'pulp-core',
               'org_id': '3340851',
               'package_count': 121,
               'snapshot': True,
               'status': 'Valid',
               'url': 'https://yum.theforeman.org/pulpcore/3.4/el7/x86_64/',
               'uuid': '5ccad0eb-56e3-4ba1-a76f-60842bb9899f'}],
     'links': {'first': '/api/content-sources/v1/repositories/?limit=100&offset=0',
               'last': '/api/content-sources/v1/repositories/?limit=100&offset=0'},
     'meta': {'count': 1, 'limit': 100, 'offset': 0}}

----------------------------------------- the next day ----------------------

          {'account_id': '0369233',
           'distribution_arch': 'any',
           'distribution_versions': ['any'],
           'failed_introspections_count': 0,
           'gpg_key': '',
           'last_introspection_error': '',
           'last_introspection_time': '2023-09-06T00:00:07Z',
           'last_snapshot': {'added_counts': {'rpm.package': 121},
                             'content_counts': {'rpm.package': 121},
                             '**created_at**': '2023-09-04T18:46:00.633897Z',
                             'removed_counts': {},
                             'repository_path': ''},
           '**last_snapshot_task_uuid**': '50e7f152-c65f-4591-8318-807762e204fc',
           '**last_snapshot_uuid**': '679906d4-091e-4d26-bf61-3586e3796a3c',
           'last_success_introspection_time': '2023-09-06T00:00:07Z',
           'last_update_introspection_time': '2023-09-04T18:46:01Z',
           'metadata_verification': False,
           'name': 'pulp-core',
           'org_id': '3340851',
           'package_count': 121,
           'snapshot': True,
           'status': 'Valid',
           'url': 'https://yum.theforeman.org/pulpcore/3.4/el7/x86_64/',
           'uuid': '5ccad0eb-56e3-4ba1-a76f-60842bb9899f'}]

**Comment by jlsherrill** [comment]
i'm going to work to add the last_snapshot_uuid updating

**Comment by jlsherrill** [comment]
closing in favor of https://github.com/content-services/content-sources-backend/pull/387

**Pull request closed without merging**